swift/extractor/SwiftExtractor.cpp

```diff
@@ -112,8 +114,21 @@ static void extractFile(const SwiftExtractorConfiguration& config,
 
 void codeql::extractSwiftFiles(const SwiftExtractorConfiguration& config,
                                swift::CompilerInstance& compiler) {
-  for (auto s : compiler.getPrimarySourceFiles()) {
-    archiveFile(config, s);
-    extractFile(config, compiler, s);
+  for (auto& [_, module] : compiler.getASTContext().getLoadedModules()) {
+    // We only extract system and builtin modules here as the other "user" modules can be built
+    // during the build process and then re-used at a later stage. In this case, we extract the
+    // user code twice: once during the module build in a form of a source file, and then as
+    // a pre-built module during building of the dependent source files.
+    if (module->isSystemModule() || module->isBuiltinModule()) {
+      llvm::SmallVector<swift::Decl*> decls;
+      module->getTopLevelDecls(decls);
+      extractDeclarations(config, compiler, module->getModuleFilename(), decls);
```

Comment:
if we add extraction of `ModuleDecl`, we could fill `decls` with just that (and we would have the module declaration itself also filled up). We can do that later, but then I'd put a TODO here maybe

Reply:
Added TODO.